src/lib.rs

```diff
@@ -1,6 +1,193 @@
+//!                         SeccompCmpArgLen::Dword,
+//!                         SeccompCmpOp::Eq,
+//!                         libc::FD_CLOEXEC as u64,
+//!                     ).unwrap(),
+//!                 ]).unwrap(),
+//!                 SeccompRule::new(vec![SeccompCondition::new(
+//!                     1,
+//!                     SeccompCmpArgLen::Dword,
+//!                     SeccompCmpOp::Eq,
+//!                     libc::F_GETFD as u64,
+//!                 ).unwrap()]).unwrap(),
+//!             ],
+//!         ),
+//!     ]
+//!     .into_iter()
+//!     .collect(),
+//!     SeccompAction::Allow,
+//!     SeccompAction::Trap,
+//!     std::env::consts::ARCH.try_into().unwrap(),
+//! ).unwrap().try_into().unwrap();
+//!
+//! seccompiler::apply_filter(&filter).unwrap();
+//! ```
+//!
+//! [`SeccompFilter`]: struct.SeccompFilter.html
+//! [`SeccompCondition`]: struct.SeccompCondition.html
+//! [`SeccompRule`]: struct.SeccompRule.html
+//! [`SeccompAction`]: struct.SeccompAction.html
+//!
+
+mod backend;
+
+use std::fmt::{Display, Formatter};
+use std::io;
+
+// Re-export the IR public types.
+pub use backend::{
+    sock_filter, BpfProgram, BpfProgramRef, Error as BackendError, SeccompAction, SeccompCmpArgLen,
+    SeccompCmpOp, SeccompCondition, SeccompFilter, SeccompRule, TargetArch,
+};
+
+// BPF structure definition for filter array.
+// See /usr/include/linux/filter.h .
+#[repr(C)]
+struct sock_fprog {
+    pub len: ::std::os::raw::c_ushort,
+    pub filter: *const sock_filter,
+}
+
+/// Library Result type.
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// Library errors.
+#[derive(Debug)]
+pub enum Error {
+    /// Error originating in the backend compiler.
+    Backend(BackendError),
+    /// Attempting to install an empty filter.
+    EmptyFilter,
+    /// System error related to calling `prctl`.
+    Prctl(io::Error),
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        use self::Error::*;
+
+        match self {
+            Backend(error) => {
+                write!(f, "Backend error: {}", error)
+            }
+            EmptyFilter => {
+                write!(f, "Cannot install empty filter.")
+            }
+            Prctl(errno) => {
+                write!(f, "Error calling `prctl`: {}", errno)
+            }
+        }
+    }
+}
+
+/// Apply a BPF filter to the calling thread.
+///
+///  # Arguments
+///
+/// * `bpf_filter` - A reference to the [`BpfProgram`] to be installed.
+///
+/// [`BpfProgram`]: struct.BpfProgram.html
+pub fn apply_filter(bpf_filter: BpfProgramRef) -> Result<()> {
+    // If the program is empty, don't install the filter.
+    if bpf_filter.is_empty() {
+        return Err(Error::EmptyFilter);
+    }
+
+    // Safe because syscall arguments are valid.
+    let rc = unsafe { libc::prctl(libc::PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0) };
+    if rc != 0 {
+        return Err(Error::Prctl(io::Error::from_raw_os_error(rc)));
+    }
```

Comment:
According to the documentation this is not the correct way of retrieving the error:
```
On error,
       -1 is returned, and errno is set to indicate the error.
```
https://man7.org/linux/man-pages/man2/prctl.2.html

Reply:
ah yes, silly mistake. I'll fix this

Reply:
fixed